os/kernel/log_dump/log_dump.c

```diff
@@ -183,54 +283,153 @@ int log_dump_save(char ch)
 
 			curbytes = 1;
 			node->arr[0] = ch;
-			sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+			sq_addlast((sq_entry_t *) node, &log_dump_chunks);
+			if (set_comp_head) {	/* new node, so compressed block starts at 0 */
+				node->comp_head = 0;
+				set_comp_head = false;
+			} else {
+				node->comp_head = -1;
+				set_comp_head = false;
+			}
 			log_dump_size += LOG_CHUNK_SIZE;
 
 		} else {
 			/* logs reached memory limit, reuse the head and free extra logs */
 			struct log_dump_chunk_s *node = (struct log_dump_chunk_s *)sq_remfirst(&log_dump_chunks);
 			curbytes = 1;
-			current_ptr = 0; /* reset the read pointer as head is modified */
+			current_ptr = 0;	/* reset the read pointer as head is modified */
 			node->arr[0] = ch;
-			sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+			sq_addlast((sq_entry_t *) node, &log_dump_chunks);
+			if (set_comp_head) {	/* new node, so compressed block starts at 0 */
+				node->comp_head = 0;
+				set_comp_head = false;
+			} else {
+				node->comp_head = -1;
+				set_comp_head = false;
+			}
 			log_dump_mem_check(max_size);
 		}
 
-	} else { /* there is place in tail */
-
+	} else {					/* there is place in tail */
+		struct log_dump_chunk_s *log_dump_tail = (struct log_dump_chunk_s *)sq_tail(&log_dump_chunks);
 		log_dump_tail->arr[curbytes] = ch;
 		curbytes += 1;
+		if (set_comp_head) {	/* new node, so compressed block starts at 0 */
+			if (log_dump_tail->comp_head < 0) {	/* compressed node not present */
+				log_dump_tail->comp_head = curbytes - 1;
+			}
+			set_comp_head = false;
+		}
 		if (curbytes % CONFIG_LOG_DUMP_MEMCHECK_SIZE == 0) {
 			log_dump_mem_check(max_size);
+			return LOG_DUMP_OK_HEAPCHK;
 		}
 	}
 
 	return LOG_DUMP_OK;
 }
 
+int log_dump_save(char ch)
+{
+	if (is_started_to_save == false) {
+		return LOG_DUMP_OK;
+	}
+
+	size_t free_size = 0;
+	int ret = 0;
+	uncomp_array[uncomp_curbytes] = ch;
+	uncomp_curbytes++;
+	if (uncomp_curbytes == CONFIG_LOG_DUMP_CHUNK_SIZE) {
+		/* compress the block, add it to the nodes, reset the uncomp_array */
+		set_comp_head = true;
+#if defined(CONFIG_LIB_LZMA)
+		out_buf = (unsigned char *)kmm_malloc(CONFIG_LOG_DUMP_CHUNK_SIZE + LZMA_PROPS_SIZE);
+#elif defined(CONFIG_LIB_MINIZ)
+		out_buf = (unsigned char *)kmm_malloc(CONFIG_LOG_DUMP_CHUNK_SIZE);
+#endif
+		if (out_buf == NULL) {
+			lldbg("memory allocation failure\n");
+			return LOG_DUMP_MEM_FAIL;
+		}
+
+		current_compression = true;
+
+		struct log_dump_msg_s msg;
+		msg.comp = false;
+		msg.ch = 'c';
+		mq_send(log_dump_get_mqfd(), (const char *)&msg, sizeof(struct log_dump_msg_s), 100);
+
+		while (current_compression) ;
+
+		ret = compress_ret;
```

Comment:
I think we don't need to assing `compress_ret` to `ret`.
We can use `compress_ret` directly, for checking with `Z_OK` and for returning `-compress_ret`, even for log.